File: main-workshop/services/dotnet/dynamic-pricing-service/Services/PricingService.cs
using Amazon.DynamoDBv2;
using Amazon.DynamoDBv2.Model;
using Amazon.EventBridge;
using Amazon.EventBridge.Model;
using System.Text.Json;

[assembly: LambdaSerializer(typeof(Amazon.Lambda.Serialization.SystemTextJson.DefaultLambdaJsonSerializer))]

public class PricingService
{
    private readonly IAmazonDynamoDB _dynamoDb;
    private readonly IAmazonEventBridge _eventBridge;
    private readonly string _tableName;
    private readonly string _eventBusName;
    private readonly JsonSerializerOptions _serializerOptions;
    private readonly Random _random;

    // Pricing constants
    private const decimal MIN_BASE_PRICE = 5.0m;
    private const decimal MAX_BASE_PRICE = 20.0m;

    public PricingService()
    {
        _dynamoDb = new AmazonDynamoDBClient();
        _eventBridge = new AmazonEventBridgeClient();
        _tableName = Environment.GetEnvironmentVariable("PRICING_TABLE_NAME") ?? "Pricing";
        _eventBusName = Environment.GetEnvironmentVariable("EVENT_BUS_NAME") ?? "";
        _random = new Random();
        
        _serializerOptions = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase
        };
    }



    public async Task<PricingResult> ProcessRideForPricing(RideCreatedEvent rideCreatedEvent, decimal rushHourMultiplier)
    {
        var result = new PricingResult
        {
            RideId = rideCreatedEvent.RideId,
            RiderId = rideCreatedEvent.RiderId
        };

        try
        {
            // Calculate price using simplified pricing rules
            var priceCalculation = CalculatePrice(rushHourMultiplier);

            // Save price calculation to database
            await SavePriceCalculationAsync(rideCreatedEvent.RideId, priceCalculation);

            // Send ride request to driver matching service
            var rideRequest = CreateRideRequest(rideCreatedEvent, priceCalculation);
            await SendPriceCalculatedEventAsync(rideRequest);

            result.Success = true;
            result.FinalPrice = priceCalculation.FinalPrice;
            result.BasePrice = priceCalculation.BasePrice;
            result.SurgeMultiplier = priceCalculation.SurgeMultiplier;
    
            return result;
        }
        catch (Exception ex)
        {
            result.Success = false;
            result.ErrorType = "UnexpectedError";
            result.ErrorMessage = ex.Message;
            throw;
        }
    }

    private PriceCalculation CalculatePrice(decimal rushHourMultiplier)
    {
        var basePrice = (decimal)(_random.NextDouble() * (double)(MAX_BASE_PRICE - MIN_BASE_PRICE) + (double)MIN_BASE_PRICE);
        basePrice = Math.Round(basePrice, 2);
        var finalPrice = Math.Round(basePrice * rushHourMultiplier, 2);

        return new PriceCalculation
        {
            BasePrice = basePrice,
            SurgeMultiplier = rushHourMultiplier,
            FinalPrice = finalPrice,
            CreatedAt = DateTime.UtcNow.ToString("O")
        };
    }

    private async Task SavePriceCalculationAsync(string rideId, PriceCalculation calculation)
    {
        var item = new Dictionary<string, AttributeValue>
        {
            { "rideId", new AttributeValue { S = rideId } },
            { "basePrice", new AttributeValue { N = calculation.BasePrice.ToString("F2") } },
            { "finalPrice", new AttributeValue { N = calculation.FinalPrice.ToString("F2") } },
            { "surgeMultiplier", new AttributeValue { N = calculation.SurgeMultiplier.ToString("F2") } },
            { "createdAt", new AttributeValue { S = calculation.CreatedAt } }
        };
        
        var putRequest = new PutItemRequest
        {
            TableName = _tableName,
            Item = item
        };

        await _dynamoDb.PutItemAsync(putRequest);
    }

    private async Task<string> SendPriceCalculatedEventAsync(RideRequest rideRequest)
    {
        if (string.IsNullOrEmpty(_eventBusName))
        {
            return "Event bus name is empty - skipping event";
        }

        var eventDetailJson = JsonSerializer.Serialize(rideRequest, _serializerOptions);

        var putEventsRequest = new PutEventsRequest
        {
            Entries = new List<PutEventsRequestEntry>
            {
                new PutEventsRequestEntry
                {
                    Source = "dynamic-pricing-service",
                    DetailType = "PriceCalculated",
                    Detail = eventDetailJson,
                    EventBusName = _eventBusName
                }
            }
        };

        var result = await _eventBridge.PutEventsAsync(putEventsRequest);
        
        var failedEntries = result.Entries.Where(e => !string.IsNullOrEmpty(e.ErrorCode)).ToList();
        if (failedEntries.Any())
        {
            throw new Exception($"Failed to send event: {failedEntries.First().ErrorCode}");
        }

        return $"Successfully sent PriceCalculated event. EventId: {result.Entries[0].EventId}";
    }

    private RideRequest CreateRideRequest(RideCreatedEvent rideEvent, PriceCalculation priceCalculation)
    {
        return new RideRequest
        {
            RideId = rideEvent.RideId,
            RiderId = rideEvent.RiderId,
            RiderName = rideEvent.RiderName,
            PickupLocation = rideEvent.PickupLocation,
            DropoffLocation = rideEvent.DestinationLocation,
            EstimatedPrice = priceCalculation.FinalPrice,
            BasePrice = priceCalculation.BasePrice,
            SurgeMultiplier = priceCalculation.SurgeMultiplier,
            PaymentMethod = rideEvent.PaymentMethod,
            Timestamp = DateTime.UtcNow,
            CorrelationId = rideEvent.CorrelationId
        };
    }

    // Event processing methods

    public async Task<PricingResult> ProcessRideCreatedEventAsync(CloudWatchEvent<RideCreatedEvent> eventBridgeEvent, decimal rushHourMultiplier)
    {
        var result = new PricingResult();

        try
        {
            var rideCreatedEvent = eventBridgeEvent.Detail;

            if (rideCreatedEvent == null)
            {
                result.Success = false;
                result.ErrorType = "DeserializationError";
                result.ErrorMessage = "Failed to deserialize ride created event";
                return result;
            }
            
            if (rideCreatedEvent.PickupLocation == null || rideCreatedEvent.DestinationLocation == null)
            {
                result.Success = false;
                result.ErrorType = "ValidationError";
                result.ErrorMessage = "Missing pickup or destination location";
                result.RideId = rideCreatedEvent.RideId;
                result.RiderId = rideCreatedEvent.RiderId;
                return result;
            }

            // Process the ride for pricing
            return await ProcessRideForPricing(rideCreatedEvent, rushHourMultiplier);
        }
        catch (JsonException ex)
        {
            result.Success = false;
            result.ErrorType = "JsonException";
            result.ErrorMessage = ex.Message;
            return result;
        }
        catch (Exception ex)
        {
            result.Success = false;
            result.ErrorType = "UnexpectedError";
            result.ErrorMessage = ex.Message;
            throw;
        }
    }
}

File: main-workshop/services/typescript/dynamic-pricing-service/src/services/PricingService.ts
import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import {
  EventBridgeClient,
  PutEventsCommand,
} from '@aws-sdk/client-eventbridge';
import { DynamoDBDocumentClient, PutCommand } from '@aws-sdk/lib-dynamodb';
import { captureAWSv3Client } from 'aws-xray-sdk-core';
import type {
  PriceCalculatedEvent,
  PriceCalculation,
  RideCreatedEvent,
} from '../models';

export interface PricingResult {
  success: boolean;
  rideId?: string;
  riderId?: string;
  finalPrice?: number;
  basePrice?: number;
  surgeMultiplier?: number;
  errorType?: string;
  errorMessage?: string;
}

export class PricingService {
  private readonly docClient: DynamoDBDocumentClient;
  private readonly tableName: string;
  private readonly eventBridge: EventBridgeClient;
  private readonly eventBusName: string;

  // Pricing constants
  private readonly MIN_BASE_PRICE = 5.0;
  private readonly MAX_BASE_PRICE = 20.0;

  constructor() {
    const client = new DynamoDBClient({});
    this.docClient = captureAWSv3Client(DynamoDBDocumentClient.from(client));
    this.tableName = process.env.PRICING_TABLE_NAME || 'Pricing';
    this.eventBridge = captureAWSv3Client(new EventBridgeClient({}));
    this.eventBusName = process.env.EVENT_BUS_NAME || '';
  }

  async processRideCreatedEvent(
    rideCreatedEvent: RideCreatedEvent,
    rushHourMultiplier: number
  ): Promise<PricingResult> {
    const result: PricingResult = {
      success: false,
    };

    try {
      if (!rideCreatedEvent) {
        result.errorType = 'DeserializationError';
        result.errorMessage = 'Failed to deserialize ride created event';
        return result;
      }

      // Validate event
      if (
        !rideCreatedEvent.pickupLocation ||
        !rideCreatedEvent.destinationLocation
      ) {
        result.errorType = 'ValidationError';
        result.errorMessage = 'Missing pickup or destination location';
        result.rideId = rideCreatedEvent.rideId;
        result.riderId = rideCreatedEvent.riderId;
        return result;
      }

      // Process the ride for pricing
      return await this.processRideForPricing(
        rideCreatedEvent,
        rushHourMultiplier
      );
    } catch (error) {
      result.errorType = 'UnexpectedError';
      result.errorMessage =
        error instanceof Error ? error.message : 'Unknown error';
      throw error;
    }
  }

  private async processRideForPricing(
    rideCreatedEvent: RideCreatedEvent,
    rushHourMultiplier: number
  ): Promise<PricingResult> {
    const result: PricingResult = {
      success: false,
      rideId: rideCreatedEvent.rideId,
      riderId: rideCreatedEvent.riderId,
    };

    try {
      // Calculate price
      const priceCalculation = this.calculatePrice(rushHourMultiplier);

      // Save price calculation to DynamoDB
      await this.savePriceCalculation(
        rideCreatedEvent.rideId,
        priceCalculation
      );

      // Publish PriceCalculated event
      await this.sendPriceCalculatedEvent(rideCreatedEvent, priceCalculation);

      result.success = true;
      result.finalPrice = priceCalculation.finalPrice;
      result.basePrice = priceCalculation.basePrice;
      result.surgeMultiplier = priceCalculation.surgeMultiplier;

      return result;
    } catch (error) {
      result.errorType = 'UnexpectedError';
      result.errorMessage =
        error instanceof Error ? error.message : 'Unknown error';
      throw error;
    }
  }

  private calculatePrice(rushHourMultiplier: number): PriceCalculation {
    // Generate random base price between MIN and MAX
    const basePrice =
      Math.random() * (this.MAX_BASE_PRICE - this.MIN_BASE_PRICE) +
      this.MIN_BASE_PRICE;
    const roundedBasePrice = Math.round(basePrice * 100) / 100;
    const finalPrice =
      Math.round(roundedBasePrice * rushHourMultiplier * 100) / 100;

    return {
      basePrice: roundedBasePrice,
      surgeMultiplier: rushHourMultiplier,
      finalPrice: finalPrice,
      createdAt: new Date().toISOString(),
    };
  }

  private async savePriceCalculation(
    rideId: string,
    calculation: PriceCalculation
  ): Promise<void> {
    const item = {
      rideId: rideId,
      basePrice: calculation.basePrice,
      finalPrice: calculation.finalPrice,
      surgeMultiplier: calculation.surgeMultiplier,
      createdAt: calculation.createdAt,
    };

    const command = new PutCommand({
      TableName: this.tableName,
      Item: item,
    });

    await this.docClient.send(command);
  }

  private async sendPriceCalculatedEvent(
    rideEvent: RideCreatedEvent,
    priceCalculation: PriceCalculation
  ): Promise<void> {
    if (!this.eventBusName) {
      return;
    }

    const priceCalculatedEvent: PriceCalculatedEvent = {
      rideId: rideEvent.rideId,
      riderId: rideEvent.riderId,
      riderName: rideEvent.riderName,
      pickupLocation: rideEvent.pickupLocation,
      dropoffLocation: rideEvent.destinationLocation,
      estimatedPrice: priceCalculation.finalPrice,
      basePrice: priceCalculation.basePrice,
      surgeMultiplier: priceCalculation.surgeMultiplier,
      paymentMethod: rideEvent.paymentMethod,
      timestamp: new Date().toISOString(),
      correlationId: rideEvent.correlationId,
    };

    const command = new PutEventsCommand({
      Entries: [
        {
          Source: 'dynamic-pricing-service',
          DetailType: 'PriceCalculated',
          Detail: JSON.stringify(priceCalculatedEvent),
          EventBusName: this.eventBusName,
        },
      ],
    });

    const result = await this.eventBridge.send(command);

    const failedEntries = result.Entries?.filter((e: any) => e.ErrorCode);
    if (failedEntries && failedEntries.length > 0) {
      throw new Error(`Failed to send event: ${failedEntries[0].ErrorCode}`);
    }
  }
}
